File: test/Authzi.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common/Initialization/RootConfiguration.fs
module RootConfiguration

open Authzi.Security
open Orleans
open System
open System.Threading.Tasks

type AccessTokenProvider() =
    let mutable accessToken = String.Empty
    member _.AccessToken with set (value) = accessToken <- value
    
    interface IAccessTokenProvider with
        member _.RetrieveTokenAsync() = Task.FromResult(accessToken);

type SiloClientProvider() =
    let mutable siloClient : IClusterClient = null
    member _.SiloClient with set (value) = siloClient <- value
    
    member _.Take() = Task.FromResult(siloClient)

let private globalAccessToken = AccessTokenProvider()
let accessTokenProvider = globalAccessToken :> IAccessTokenProvider

let siloClientProvider = SiloClientProvider()

let getClusterClient (accessToken: string) =
    globalAccessToken.AccessToken <- accessToken
    siloClientProvider.Take().Result

File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common.CSharp/Initialization/RootConfiguration.cs
using System;
using System.Threading.Tasks;
using AuthZI.Security;
using AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common;
using Orleans;

public static class RootConfiguration
{
  private static readonly ConfigurableAccessTokenProvider GlobalAccessToken = new();

  public static IAccessTokenProvider accessTokenProvider => GlobalAccessToken;

  public static IClusterClient getClusterClient(string accessToken)
  {
    GlobalAccessToken.AccessToken = accessToken;

    return TestData.IClusterClient;
  }

  private sealed class ConfigurableAccessTokenProvider : IAccessTokenProvider
  {
    public string AccessToken { private get; set; } = string.Empty;

    public Task<string> RetrieveTokenAsync() => Task.FromResult(AccessToken);
  }
}
